File: sample_code/variables.cpp
#include "variables.h"
#include <sstream>
#include <iostream>
using namespace std;

VariableTracker::VariableTracker() {
	locationEnd = 0;
	robotEnd = 0;
	craneEnd = 0;
	pileEnd = 0;
	containerEnd = 0;
	groundEnd = 0;
}

	// Construct a literal / variable tracker
VariableTracker::VariableTracker(int numLocs, int numRobots, int numCranes,
	int numPiles, int numContainers) {
	locationEnd = numLocs;
	robotEnd = numRobots+locationEnd;
	craneEnd = numCranes + robotEnd;
	pileEnd = numPiles + craneEnd;
	containerEnd = numContainers + pileEnd;
	groundEnd = containerEnd + 1;
}

// Get the unique integer id associated with name of variable or literal
int VariableTracker::getId(string var) const {
	istringstream instr(var);
	char ch;
	int num;
	instr >> ch >> num;
	switch(ch){
		case 'l': //location
			return num;
		case 'r': //robot
			return locationEnd+num;
		case 'k': //crane
			return robotEnd+num;
		case 'p': //pile
			return craneEnd+num;
		case 'c': //container
			return pileEnd+num;
		case 'G': //ground
			return containerEnd;
		case 'x': //variable
			return groundEnd+num;
		default:
			cerr << "Invalid literal\n";
			return -1;
	};
}

// Get the name of the variable or literal associated with id
string VariableTracker::getName(int n) const {
	ostringstream varName;
	if (n < 0) varName << "InvalidVariable";
	else if (n < locationEnd) varName << 'l' << n;
	else if (n < robotEnd) varName << 'r' << n-locationEnd;
	else if (n < craneEnd) varName << 'k' << n-robotEnd;
	else if (n < pileEnd) varName << 'p' << n-craneEnd;
	else if (n < containerEnd) varName << 'c' << n-pileEnd;
	else if (n < groundEnd) varName << 'G';
	else varName << 'x' << n-groundEnd;
	return varName.str();
}

File: variables.py
'''
variables.py
------------
Implements a class that keeps track of variable names
'''

'''
VariableTracker
---------------
In our model, all variables and ground literals are uniquely
represented as integers.

VariableTracker provides an interface that maps back-and-forth
between variable names like "c0, p1" and their unique integer identifiers

It works like this:
All integers x
0 <= x <= numLocs+numRobots+numCranes+numPiles+numContainers+1
are ground literals.
Any higher integers are variables. Integers below 0 are invalid.

If for example, we have
locationEnd <= x < robotEnd
then we know that x is a robot r with number = x - locationEnd
So locationEnd = 4, robotEnd = 15, then integer 11 represents the robot r7

All you need to know is how to construct a variable tracker 
and use its provided functions.
'''

class VariableTracker:

    ## Construct a VariableTracker object
    def __init__(self, numLocs=0, numRobots=0, numCranes=0, numPiles=0, numContainers=0):
        self.locationEnd = numLocs
        self.robotEnd = numRobots + self.locationEnd
        self.craneEnd = numCranes + self.robotEnd
        self.pileEnd = numPiles + self.craneEnd
        self.containerEnd = numContainers + self.pileEnd
        self.groundEnd = self.containerEnd + 1

    ## Get the unique integer id of a variable or literal
    def getId(self, var):
        ch = var[0]
        if ( ch != 'G' ):
            num = int(var[1])

        if ch == 'l': ## location
            return num
        elif ch == 'r': ## robot
            return self.locationEnd + num
        elif ch == 'k': ## crane
            return self.robotEnd + num
        elif ch == 'p': ## pile
            return self.craneEnd + num
        elif ch == 'c': ## container
            return self.pileEnd + num
        elif ch == 'G': ## ground
            return self.containerEnd
        elif ch == 'x': ## variable
            return self.groundEnd + num
        else:
            print "Invalid literal\n"
            return -1


    ## Get the name of the variable associated with id (n)
    def getName(self, n):
        varName = ''
        if n < 0: varName = "InvalidVariable"
        elif n < self.locationEnd: varName = 'l' + str(n)
        elif n < self.robotEnd: varName = 'r' + str(n - self.locationEnd)
        elif n < self.craneEnd: varName = 'k' + str(n - self.robotEnd)
        elif n < self.pileEnd: varName = 'p' + str(n - self.craneEnd)
        elif n < self.containerEnd: varName = 'c' + str(n - self.pileEnd)
        elif n < self.groundEnd: varName = 'G'
        else: varName = 'x' + str(n - self.groundEnd)
        return varName


    ## Get the integer value of the first variable
    def getFirstVar(self):
        return self.groundEnd

    ## Returns true if id (n) is a valid literal or variable
    def isValid(self, n):
        return (n >= 0)

    ## Returns true if v is a variable
    def isVariable(self, v):
        return (v >= self.groundEnd)

    ## Returns true if l is a literal
    def isLiteral(self, l):
        return (l < self.groundEnd and l >= 0)



